### ISSUE TITLE
Boolean columns are dropped with st_write_db

### ISSUE THREAD

**Issue by dpprdan** [r-spatial/sf#592]
`st_write_db` drops boolean/logical columns, see reprex below. I [just learned](https://github.com/r-spatial/sf/issues/579#issuecomment-348728236) that this makes sense with `st_write`, because GDAL does [not support](http://www.gdal.org/ogr__core_8h.html#a787194bea637faf12d61643124a7c9fc) boolean field types. However, this is not the case with (R)PostgreSQL.

Currently I also don't see a workaround (e.g. manually defining PostgreSQL field types).

(Is `clean_columns` even necessary with `st_write_db`? AFAIK RPostgreSQL handles the conversion of factors and non-UTF-8 character strings, but not sure regarding POSIXlt.).

<details><summary>Reprex & Session info</summary>

``` r
library("RPostgreSQL")
#> Lade nötiges Paket: DBI
library("sf")
#> Linking to GEOS 3.6.1, GDAL 2.2.0, proj.4 4.9.3

pgcon <- 
  dbConnect(
    PostgreSQL(),
    host = "localhost",
    dbname = "postgres"
  )

nc_bool <- st_read(system.file("shape/nc.shp", package="sf"))
#> Reading layer `nc' from data source `C:\Users\daniel\Documents\R\win-library\3.4\sf\shape\nc.shp' using driver `ESRI Shapefile'
#> Simple feature collection with 100 features and 14 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
#> epsg (SRID):    4267
#> proj4string:    +proj=longlat +datum=NAD27 +no_defs
nc_bool$large_area <- nc_bool$AREA >= 0.15

st_write_db(pgcon, nc_bool)
#> ignoring columns with unsupported type:
#> [1] "large_area"

dbDisconnect(pgcon)
#> [1] TRUE
```

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.2 (2017-09-28)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  German_Germany.1252         
#>  tz       Europe/Berlin               
#>  date     2017-12-12
#> Packages -----------------------------------------------------------------
#>  package     * version date       source                               
#>  backports     1.1.1   2017-09-25 CRAN (R 3.4.1)                       
#>  base        * 3.4.2   2017-09-28 local                                
#>  compiler      3.4.2   2017-09-28 local                                
#>  datasets    * 3.4.2   2017-09-28 local                                
#>  DBI         * 0.7     2017-06-18 CRAN (R 3.4.0)                       
#>  devtools      1.13.4  2017-11-09 CRAN (R 3.4.2)                       
#>  digest        0.6.12  2017-01-27 CRAN (R 3.4.2)                       
#>  evaluate      0.10.1  2017-06-24 CRAN (R 3.4.0)                       
#>  graphics    * 3.4.2   2017-09-28 local                                
#>  grDevices   * 3.4.2   2017-09-28 local                                
#>  grid          3.4.2   2017-09-28 local                                
#>  htmltools     0.3.6   2017-04-28 CRAN (R 3.4.0)                       
#>  knitr         1.17    2017-08-10 CRAN (R 3.4.1)                       
#>  magrittr      1.5     2014-11-22 CRAN (R 3.4.2)                       
#>  memoise       1.1.0   2017-10-16 Github (hadley/memoise@d63ae9c)      
#>  methods     * 3.4.2   2017-09-28 local                                
#>  Rcpp          0.12.14 2017-11-23 CRAN (R 3.4.2)                       
#>  rmarkdown     1.8     2017-11-17 CRAN (R 3.4.2)                       
#>  RPostgreSQL * 0.6-3   2017-11-16 Github (tomoakin/RPostgreSQL@bfdb284)
#>  rprojroot     1.2     2017-01-16 CRAN (R 3.4.2)                       
#>  sf          * 0.5-5   2017-10-06 Github (r-spatial/sf@5f7a415)        
#>  stats       * 3.4.2   2017-09-28 local                                
#>  stringi       1.1.6   2017-11-17 CRAN (R 3.4.2)                       
#>  stringr       1.2.0   2017-02-18 CRAN (R 3.4.2)                       
#>  tools         3.4.2   2017-09-28 local                                
#>  udunits2      0.13    2016-11-17 CRAN (R 3.4.1)                       
#>  units         0.4-6   2017-08-27 CRAN (R 3.4.1)                       
#>  utils       * 3.4.2   2017-09-28 local                                
#>  withr         2.1.0   2017-11-01 CRAN (R 3.4.2)                       
#>  yaml          2.1.15  2017-12-01 CRAN (R 3.4.3)
```

</details>

**Comment by dpprdan** [comment]
This issue seems to be solved by \#558.

In case someone needs a workaround until this is on CRAN: convert the logical column to character and define the [field.type](https://www.postgresql.org/docs/current/static/datatype.html) (for all columns) manually.

``` r
library("RPostgreSQL")
#> Lade nötiges Paket: DBI
library("sf")
#> Linking to GEOS 3.6.1, GDAL 2.2.0, proj.4 4.9.3

pgcon <- 
  dbConnect(
    PostgreSQL(),
    host = "localhost",
    dbname = "postgis"
  )

nc_bool <- st_read(system.file("shape/nc.shp", package="sf"))
#> Reading layer `nc' from data source `C:\Users\daniel\Documents\R\win-library\3.4\sf\shape\nc.shp' using driver `ESRI Shapefile'
#> Simple feature collection with 100 features and 14 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
#> epsg (SRID):    4267
#> proj4string:    +proj=longlat +datum=NAD27 +no_defs
nc_bool$large_area <- nc_bool$AREA >= 0.15

# too lazy to define field.types for all columns, so picking only the boolean
# and geometry
nc_bool <- nc_bool[c("large_area", "geometry")]

# convert logical/boolean to character
nc_bool$large_area <- as.character(nc_bool$large_area)

# define field types
field_types_pg <-
  c(
    "large_area"   = "boolean",
    "geometry"     = "geometry(MultiPolygon,4267)" # modify geometry type and crs accordingly
  )


st_write_db(pgcon, nc_bool, row.names = FALSE, field.types = field_types_pg)

dbDisconnect(pgcon)
#> [1] TRUE
```

**Comment by edzer** [comment]
We're in the process of integrating `st_write_db` into `st_write`, see https://github.com/r-spatial/sf/pull/558; I guess we need to revisit `clean_columns` to have it accept cases (like boolean) that RPostgreSQL and/or DBI accept, but GDAL doesn't. Am I right? @etiennebr

**Comment by dpprdan** [comment]
Yes, I have seen that and as far as [I can tell](https://github.com/r-spatial/sf/pull/558#issuecomment-351327384), @etiennebr already took care of this. I.e. `st_write` does not use `clean_columns` when writing to db anymore.

**Comment by etiennebr** [comment]
Yes, I wasn't sure what to do about `clean_column`. I manually added the `units` (and `sfc`) conversion in `dbDataType()` to rely on `DBI` as much as possible, but there might be other corner cases that I missed. As @dpprdan mentioned, each database has it's own way of handling different types, so it should probably be left to each database driver as much as possible.

**Comment by edzer** [comment]
Can we close this here?

**Comment by dpprdan** [comment]
AFAIK it's not fixed in `master` until #558 is merged. Not sure, though, how you usually close issues, i.e. is the existence of a fix in a pull request enough, even though it's not merged yet?

**Comment by edzer** [comment]
Thanks; that still fails for me with
```r
> test_check("sf")
1. Failure: can read from db (@test_postgis_ODBC.R#119) ------------------------
error$message does not match "no applicable method for 'st_read'".
Actual value: "dsn should specify a data source or filename"


2. Failure: new SRIDs are handled correctly (@test_postgis_ODBC.R#271) ---------
st_crs(x) == crs isn't true.


3. Failure: can read from db (@test_postgis_RPostgreSQL.R#119) -----------------
error$message does not match "no applicable method for 'st_read'".
Actual value: "dsn should specify a data source or filename"

```

**Comment by edzer** [comment]
@etiennebr I'm close to submitting 0.6-1 to CRAN; shall we leave this resting to the next release opportunity (once every two months)?

**Comment by etiennebr** [comment]
How close? I can finish it by the end of the week.

**Comment by edzer** [comment]
That would be great. 0.6-1 is needed for a CRAN release of `stars`; I bundled all GDAL stuff in `sf`.

**Comment by edzer** [comment]
I will submit 0.6-1 on Monday.

**Comment by etiennebr** [comment]
@edzer, I'm deeply sorry, but I can't finish it for Monday; I still have issues with multi-geom columns and I don't want to rush it. I think it will also be a good thing to test the new features on the dev version before the CRAN release.

**Comment by edzer** [comment]
sure, no worries!

**Comment by etiennebr** [comment]
Thanks for the feedback @dpprdan. I can reproduce the `WKB buffer too small` error with `odbc`. 

I think you're right, there should be a warning when the driver drops a column. Could you open another issue for that as well?

**Comment by edzer** [comment]
Looks like something I caused. A `traceback()` right after the error message would be helpful!

**Comment by dpprdan** [comment]
Closing this since the original problem was solved by #558 and #721 and #722 were opened (with #722 already solved) for the related problems with odbc and GDAL.
